docs: add more thorough examples for predicates (#2345)
FuelLabs/fuels-ts: apps/docs/src/guide/predicates/methods.md

# Interacting With Predicates

There are 4 ways to interact with predicates: `populateTransactionPredicateData`, `sendTransaction`, `simulateTransaction`, `getResourcesToSpend`

> ## `populateTransactionPredicateData`

Reviewer: This isn't strictly true as a `Predicate` extends an `Account` so it inherits all the account based functionality (submitting, querying etc).

Maybe this could be made more general with the subheadings:
- Checking Balances - `getBalances` & `getResourcesToSpend` (make this one tx specific)
- Transactions -`sendTransaction` / `simulateTransaction` and `populateTransactionPredicateData` (for where the predicate is added to a tx builder) 
- Transfers - `createTransfer` / `transfer`

FYI I don't necessarily agree with this design choice as there are some things in there that don't make sense i.e. signing and the use of connectors, but we should highlight the account based functionality that it does use.

# Interacting With Predicates

> The `Predicate` class extends the [`Account`](https://docs.fuel.network/docs/fuels-ts/account/) class, and therefore it inherits all of it's methods.  Therefore, there are multiple ways to interact with predicates, but broadly speaking there are 3 general use cases that would be of interest to a user. We can think of them in terms of `Checking Balances`, `Transactions` and `Transfers`

Reviewer 2: ```suggestion
The `Predicate` class extends the [`Account`](https://docs.fuel.network/docs/fuels-ts/account/) class, inheriting all its methods. Therefore, there are multiple ways to interact with predicates, but broadly speaking, we can think about three:
 - `Checking Balances`
 - `Transactions`
 - `Transfers`
```

## Checking Balances

### `getBalances`

This will return the balances of the wallet that is being used to interact with the predicate.

<<< @/../../docs-snippets/src/guide/wallets/checking-balances.test.ts#checking-balances-2{ts:line-numbers}

### `getResourcesToSpend`

> This will return the resources that will be spent when calling the predicate. This is useful when creating the predicate transaction so that you can ensure that the wallet has sufficient funds to cover the transaction fees.

Reviewer 2: ```suggestion
This will return the resources that are spent when calling the predicate. This is useful when creating the predicate transaction to ensure that the wallet has sufficient funds to cover the transaction fees.
```

> This will return the resources that will be spent when calling the predicate. This is useful when creating the predicate transaction so that you can ensure that the wallet has sufficient funds to cover the transaction fees.

Author: https://github.com/FuelLabs/fuels-ts/commit/ee69754a2d64947dcb59dc454b35a6cbeed52108

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-1{ts:line-numbers}

## Transactions

> ## `sendTransaction`

Author: I intentionally made these sub-headings of the broader headings of  `Checking Balances`, `Transactions` and `Transfers` as I think that categorizes these methods more precisely

> ## `sendTransaction`

Author: My bad, I misread your suggestion 😅  I agree with you, thanks for pointing this out.

> Once you've populated the transaction with the populated data you can send the transaction to the node using the `sendTransaction` method.

Reviewer 2: ```suggestion
Once you've populated the transaction with the populated data, you can send it to the node using the `sendTransaction` method.
```

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-2{ts:line-numbers}

> ## `simulateTransaction`

Author: See https://github.com/FuelLabs/fuels-ts/pull/2345#discussion_r1609613466

The `simulateTransaction` method should be used to dry-run a predicate call, ensuring that the wallet used has sufficient funds to cover the transaction fees, without consuming any resources.

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-3{ts:line-numbers}

### `populateTransactionPredicateData`

> Once you've instantiated a `new Predicate` you will want to create a transaction request containing the `gasLimit` and `maxFee` then pass this to the `populateTransactionPredicateData` method which will populate the transaction with the required data to be sent to the node.

Reviewer 2: ```suggestion
Once you've instantiated a `new Predicate` you will want to create a transaction request containing the `gasLimit` and `maxFee` and then pass this to the `populateTransactionPredicateData` method, which will populate the transaction with the required data to be sent to the node.
```

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-4{ts:line-numbers}

## Transfers

### `createTransfer`

> The `createTransfer` method is used to create a transfer object that can be used to send funds to another wallet. This can be useful should you need to modify the transfer object or use it's properties before sending it to the node.

Author: ee69754

<<< @/../../docs-snippets/src/guide/predicates/send-and-spend-funds-from-predicates.test.ts#send-and-spend-funds-from-predicates-8{ts:line-numbers}

### `transfer`

> Once you've created a `transfer` object using the `createTransfer` method can send the funds to another wallet using the `transfer` method.

Author: https://github.com/FuelLabs/fuels-ts/commit/ee69754a2d64947dcb59dc454b35a6cbeed52108

# Interacting With Predicates

The `Predicate` class extends the [`Account`](https://docs.fuel.network/docs/fuels-ts/account/) class, inheriting all its methods. Therefore, there are multiple ways to interact with predicates, but broadly speaking, we can think about three:
 - `Checking Balances`
 - `Transactions`
 - `Transfers`

## Checking Balances

> 
> ### `getBalances`
> 
> This will return the balances of all assets owned by the predicate.
> 
> See also: [Checking Wallet Balances](https://docs.fuel.network/docs/fuels-ts/wallets/checking-balances/#getting-a-wallets-balance)

Reviewer 2: I realized my first comment may have been misleading.

What I meant was to remove this section entirely, and then — see the next suggestion.

```suggestion
```

Reviewer 2: Dismissing my own suggestion. 🤦‍♂️

### `getResourcesToSpend`

This will return the resources owned by a predicate so that they can be added to a transaction request.

This method is called under the hood when using [`transfer`](./methods.md#transfer) or [`createTransfer`](./methods.md#createtransfer).

You may want to use this method when using a predicate in an existing transaction request.

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-1{ts:line-numbers}

> 

Reviewer 2: Then the "see also" would go below this snippet here.

However, considering there is no `getBalances` for `Predicates`, why should we have this here?

```suggestion

See also:
 - [Checking Wallet Balances](https://docs.fuel.network/docs/fuels-ts/wallets/checking-balances/#getting-a-wallets-balance)
```

Reviewer 2: Never mind.

Since `Predicates` extends from `Account`, there is also a `getBalances` for it (thanks, @Torres-ssf, for the heads-up).

With that in mind, forget everything I said. 🤦‍♂️ You could leave it as is or roll it back to your first version, as that would be more complete and remove the need for users to click around.

Perhaps the phrasing should target `Account` instead of `Wallets`, though. I'm unsure if that makes sense, but it can also be a follow-up. Sorry for the confusion.

Author: No worries, it was also a bit misleading for me before I understood that a predicate is a type of account at first. I think the current state makes sense and still references another example of getting balances that consumers may also be familiar with.

## Transactions

### `sendTransaction`

This is used to send a transaction to the node.

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-2{ts:line-numbers}

### `simulateTransaction`

The `simulateTransaction` method should be used to dry-run a predicate call, ensuring that there are sufficient funds to cover the transaction fees, without consuming any resources.

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-3{ts:line-numbers}

## Transfers

### `createTransfer`

> The `createTransfer` method creates a transfer request and add the transfer details to that request. This method estimates the transaction cost via a dry-run, and then funds it with predicate resources, which is useful since the returned transaction request can now be submitted with greater certainty that it will succeed.
> 
>  This can be useful should you need to modify the transfer request or use its properties before sending it to the node.

Reviewer 2: ```suggestion
The `createTransfer` method creates a transaction request with all the necessary transfer details. It automatically estimates the transaction costs via a dry-run call and funds the request with the required predicate resources. After this, one can submit the returned transaction request with greater certainty that it will succeed.

However, please remember that you can still modify the transfer request details and use its properties before submitting it to the node.
```

I'd also say these two lines seem irrelevant to the purpose of the code snippet and could be deleted:

```ts
const chainId = provider.getChainId();

const txId = transactionRequest.getTransactionId(chainId);
```

Author: My apologies, when I applied the suggestion you made [here](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/e4e78bb58cd57c741b85d1273d7960aa0f2d3a40) Github automatically resolved the issue, and so that consideration eluded me. 

This code snippet is re-used from another [guide](https://github.com/FuelLabs/fuels-ts/blob/mc/docs/improve-predicate-docs/apps/docs/src/guide/predicates/send-and-spend-funds-from-predicates.md#L45), but I agree with your point, I'll implement a new snippet for this.

Reviewer 2: Perhaps you can just move the two lines below where the snippet ends.

Author: See [`2a8de44` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/2a8de44dc130234cfff5cc6b455c9ad6f1be188f)

Reviewer 2: TBH, the text preceding this snippet could be clearer as well. 🙂 

We could move the last line of text inside the snippet as a comment.

Both cases would benefit from it, and we could share the same snippet.

See what you think.

![suggestion](https://github.com/FuelLabs/fuels-ts/assets/26660/688a8681-2870-41db-a457-d3f725a0cf73)

Reviewer 2: Ooof, I didn't see your message before sending mine.

The UI may not refresh while one is writing a reply. 😓

Author: > Ooof, I didn't see your message before sending mine.
> 
> The UI may not refresh while one is writing a reply. 😓

Happens to me quite frequently as well, I end up missing comments at times 😢 

> TBH, the text preceding this snippet could be clearer as well. 🙂
> 
> We could move the last line of text inside the snippet as a comment.
> 
> Both cases would benefit from it, and we could share the same snippet.
> 
> See what you think.
> 

I agree, and it promotes code reuse so I will make that change.

Author: See [`80af938` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80af938e61d654730c281f02f729725b79f97b10)

Reviewer 2: Nice! Not sure, but then I guess you'll need to undo [this](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/2a8de44dc130234cfff5cc6b455c9ad6f1be188f#diff-bf9cefcebbc4d0127bda5f3a965bcb75c3d6e4130467325e4701ed38d0b91dd1R122-R156) as well I?

Anyway, I am leaving this to you and resolving it for now.

Author: Correct 😅

# Interacting With Predicates

The `Predicate` class extends the [`Account`](https://docs.fuel.network/docs/fuels-ts/account/) class, inheriting all its methods. Therefore, there are multiple ways to interact with predicates, but broadly speaking, we can think about three:
 - `Checking Balances`
 - `Transactions`
 - `Transfers`

## Checking Balances

### `getBalances`

> This will return the balances of the wallet that is being used to interact with the predicate.

Author: [`80b13c2` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a)

> <<< @/../../docs-snippets/src/guide/wallets/checking-balances.test.ts#checking-balances-2{ts:line-numbers}

Reviewer 3: The snippet is in `wallets/...` and it should be in `predicates/...`.

Author: It's being reused for this purpose but it was originally uses by the wallets documentation.

Author: https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a

Reviewer 2: Un-resolving this just to ask: What was the resolution?

Author: I must have overlooked this comment, my bad. I agree with your suggestion.

Author: See [`2a8de44` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/2a8de44dc130234cfff5cc6b455c9ad6f1be188f)

### `getResourcesToSpend`

This will return the resources owned by a predicate so that they can be added to a transaction request.

> This method is called under the hood when using `transfer` or `createTransfer`.

Author: [`80b13c2` (#2345)](https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a)

You may want to use this method when using a predicate in an existing transaction request.

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-1{ts:line-numbers}

## Transactions

### `sendTransaction`

> Once you've populated the transaction with the populated data, you can send it to the node using the `sendTransaction` method.

Author: https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a

<<< @/../../docs-snippets/src/guide/predicates/interacting-with-predicates.test.ts#interacting-with-predicates-2{ts:line-numbers}

### `simulateTransaction`

> The `simulateTransaction` method should be used to dry-run a predicate call, ensuring that the wallet used has sufficient funds to cover the transaction fees, without consuming any resources.

Author: https://github.com/FuelLabs/fuels-ts/pull/2345/commits/80b13c26713a2e33bb6a64d252796261e83a7f1a

### `getBalances`

> This will return the balances of the account that owns the resources in the predicate.

Reviewer 3: ```suggestion
This will return the balances of all assets owned by the predicate.
```
The predicate itself owns the resources and _it_ is the account.
